---
title: "cfdnakit vignette"
author: "Pitithat Puranachot"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{cfdnakit vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center"
)
```

This vignette demonstrates how cfdnakit package works. First, load cfdnakit package to environment

```{r setup}
library(cfdnakit)
```


Let cfdnakit read sequence alignment file (.bam) with function read_bamfile. An BAM index file (.bai) is necessary for efficiently reading the file. If it doesn't exist in the same directory, this function will automatically create one. This function will split sequence reads into equal-size non-overlapping windows. Available size of bin are 100, 500, and 1000 KB. A path to a bam file is given to the function read_bamfile. sample_bins = read_bamfile(bamfile_path = "Patient1.bam", binsize = 1000) For demonstration, we load instance of an example sequence from file example_patientcfDNA.RDS. We do not filtering out reads in "Blacklist Regions" in this example.

```{r read a bam file and splited into 1000 KB non-overlapping bins, warning=FALSE}

sample_bamfile = system.file("extdata","ex.plasma.bam",
                package = "cfdnakit")
plasma_SampleBam = read_bamfile(sample_bamfile,
                                apply_blacklist = FALSE)
```

This will take a period of time depending on the size of your BAM file. We recommend to save the result as RDS file using saveRDS function.

```
saveRDS(plasma_SampleBam, file = patientcfDNA_SampleBam.RDS)
```

This vignette will use the example data (example_patientcfDNA.RDS). Now we load this file into R session.

```{r load example patient cfDNA sample BAM, warning=FALSE}

example_RDS="example_patientcfDNA.RDS"
example_RDS_file =
    system.file("extdata",example_RDS,
                package = "cfdnakit")
sample_bins = readRDS(example_RDS_file)

```

Fragment-length distribution of the sample can be visualized with function plot_fragment_dist. The most common fragment-length is shown at the top-right legend. The x-axis is the length of cfDNA; y-axis is the proportion of cfDNA fragment having a specific length.

```{r plot fragment length distribution, fig.width=5,fig.height=3.5}
plot_fragment_dist(list("Patient.1"=sample_bins))
```


We want to compare the fragment length distribution of patient's cfDNA with that of a healthy individual. For demonstration purpose, we derived a cfDNA from Snyder et al. (2016) and create a RData file "BH01_chr15.RDS". This file can be loaded in R environment with readRDS function.

```{r getting build-in healthy plasma cfDNA sample, warning=FALSE}
control_rds="BH01_chr15.RDS"
control_RDS_file =
    system.file("extdata",control_rds,
                package = "cfdnakit")
control_bins =
    readRDS(control_RDS_file)

```


Create a list of sample bam files and plot their distribution together with function plot_fragment_dist. The list must be given a distinct sample name (e.g. Healthy.cfDNA) for each sample. 

```{r plot fragment length distribution comparing to a healthy cfDNA, fig.width=5.5, fig.height=4}
comparing_list = list("Healthy.cfDNA"=control_bins,
                      "Patient.1"=sample_bins)
plot_fragment_dist(comparing_list)

```
Comparing multiple cfDNA samples is possible by creating list of sample_bins; each with unique sample name. 


We can extract genome-wide fragment-length profile. 

```{r getting fragment length profile,warning=FALSE, fig.width=12, fig.height=5}
sample_profile = 
  get_fragment_profile(sample_bins,
                       sample_id = "Patient.1")
sample_profile$sample_profile
# plot_isize_distribution(sample_profile)
```


We can plot genome-wide short-fragment ratio with the function plot_sl_ratio. Given short-fragment profile, short-fragment ratio per bin infer contribution of tumor-derived fragment (ctDNA) into cfDNA. The enrichment of short-fragment cfDNA in a large genomic region could infer the copy-number aberration status. The higher short-fragment ratio indicate amplification event whereas deletion would have relatively lower short-fragment ratio. It is base-on a hypothesis that a tumor cell without CNV contributes its DNA equally throughout the genome into the circulation.

```{r plot genome-wide short-fragment ratio, fig.width=12, fig.height=5, warning=FALSE}
plot_sl_ratio(sample_profile,ylim = c(0.25,1.75))
```



Save fragment profile as RDS file for later use or for creation of Panel-of-Normal for further analysis

```{r saving profile as RDS}
package_dir = system.file("extdata", package = "cfdnakit")
saveRDS(sample_profile,
        file = paste0(package_dir,"/plasma.example.RDS"))

```

Making a Panel-of-Normal is necessary for downstream analysis as we want to compare fragment profile between a cfDNA from patient with pooled of healthy individuals. First, we create a text file where each line is a full path to rds file created by aforementioned function.

```{r Make a file containing path to fragment profile}
path_to_PoN_txt = 
  system.file("extdata", "reference_healthy.listfile",
              package = "cfdnakit")
pon_files = readLines(path_to_PoN_txt)
head(pon_files, n=5)
# 
```

create_PoN function will read through all fragment profile files and create a set of PoN profile. We save this PoN profile and load into analysis later.
```
PoN.profiles = create_PoN(path_to_PoN_txt)
saveRDS(PoN.profiles, "PoN.rds")
```

For this vignette, we load ready-to-use PoN profile (ex.PoN.rds). 
```{r Make a pon profile}
PoN_rdsfile = system.file("extdata",
                          "ex.PoN.rds",
                          package = "cfdnakit")
PoN.profiles = readRDS(PoN_rdsfile)

```


We rescale the SLRatio by the median of PoN samples. 

```{r Reading PoN samples and plot scaled value,warning=FALSE,fig.width=8, fig.height=4}
sample_zscore = 
  get_zscore_profile(sample_profile,
                     PoN.profiles)
sample_zscore_segment = segmentByPSCB(sample_zscore)
plot_transformed_sl(sample_zscore,sample_zscore_segment)
```


Performing CNV Calling and plot distance matrix.
```{r cnv calling and plot distance matrix,warning=FALSE,fig.width=6, fig.height=4}

sample_cnv = call_cnv(sample_zscore_segment,sample_zscore)

plot_distance_matrix(sample_cnv)

solution_table = get_solution_table(sample_cnv)

solution_table
```

We select a solution and plot the CNV Calling result.
```{r plot cnv-calling first solution,warning=FALSE,fig.width=12, fig.height=5}
plot_cnv_solution(sample_cnv,selected_solution = 1)

```

Calculate CPA Score from the segmentation result
```{r calculate CPA score}
calculate_CPA_score(sample_zscore_segment)
```
