---
title: "cfDNAKit"
author: "Pitithat"
date: "5/13/2021"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(dev = 'pdf')

```

This rmarkdown to demonstrate how cfdnakit work. First, load cfdnakit package to environment

```{r load cfdnakit package}
library(cfdnakit)
```

Let cfdnakit read sequence alignment file (.bam) with function read_bamfile. This function will split sequence reads into equal-size non-overlapping windows. Available size of bin are 100, 500, and 1000 KB.

```{r read a bam file and splited into 1000 KB non-overlapping bins}

sample_bins = 
  read_bamfile(bamfile_path = "/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/results_per_pid/OE0290-PED_5LB-017/alignment_umi/plasma-01-01_OE0290-PED_5LB-017_merged.mdup_groupbyUMI_callConsensus_realigned.bam",
               binsize = 1000)

paste0("Number of bin : ",length(sample_bins))

```

Extract cfDNA fragment length information.

```{r getting fragment length profile distribution,warning=FALSE, fig.width=6, fig.height=5}
sample_profile = 
  get_fragment_profile(sample_bins,
                       sample_id = "plasma-01-01_OE0290-PED_5LB-017")
plot_isize_distribution(sample_profile)
```



```{r plot genome-wide SLRatio, warning=FALSE, fig.width=7, fig.height=3}
plot_sl_ratio(sample_profile)
```

Save fragment profile as RDS file for later use or for creating Panel-of-Normal

```{r saving profile as RDS}
save_fragment_profile(sample_profile,
                      output_dir = "/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/results_per_pid/OE0290-PED_5LB-017/cfdnakit/plasma-01-01_OE0290-PED_5LB-017/")

```

Making a Panel-of-Normal is necessary for downstream analysis as we want to compare fragment profile between a cfDNA from patient with pooled of healthy individuals. First, we create a text file where each line is a full path to rds file created by aforementioned function.

```{r Make a pon profile from a text file containing list of rds}
path_to_PoN_txt = "/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/processing_scripts/cfdnakit_script/reference_healthy.listfile"

create_PoN(path_to_PoN_txt,output_dir = "/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/processing_scripts/cfdnakit_script/")


```

We rescale the SLRatio by the median of PoN samples. 

```{r Reading-in Panel-of-Normal samples and plot scaled value,warning=FALSE}
PoN_rdsfile = "/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/processing_scripts/cfdnakit_script/PoN_fragment_profiles.rds"
sample_zscore = get_zscore_profile(sample_profile,PoN_rdsfile)
sample_zscore_segment = segmentByPSCB(sample_zscore)
plot_transformed_sl(sample_zscore,sample_zscore_segment)
```


Performing CNV Calling and plot distance matrix.
```{r cnv calling and plot distance matrix,warning=FALSE,fig.width=7, fig.height=5}

sample_cnv = call_cnv(sample_zscore_segment,sample_zscore)

plot_distance_matrix(sample_cnv)

solution_table = get_solution_table(sample_cnv)

solution_table
```

We select a solution and plot the CNV Calling result.
```{r plot cnv-calling first solution,warning=FALSE,fig.width=7, fig.height=3}
plot_cnv_solution(sample_cnv,selected_solution = 1)

```

Calculate PCA Score from the segmentation result
```{r calculate PCA score}
calculate_PCA_score(sample_zscore_segment)
```
