<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Pitithat" />


<title>Demonstration cfdnakit</title>

<script src="demonstration_markdown_files/header-attrs-2.7/header-attrs.js"></script>
<script src="demonstration_markdown_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="demonstration_markdown_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="demonstration_markdown_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="demonstration_markdown_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="demonstration_markdown_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="demonstration_markdown_files/navigation-1.1/tabsets.js"></script>
<link href="demonstration_markdown_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="demonstration_markdown_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Demonstration cfdnakit</h1>
<h4 class="author">Pitithat</h4>
<h4 class="date">5/13/2021</h4>

</div>


<p>This rmarkdown to demonstrate how cfdnakit work. First, load cfdnakit package to environment</p>
<pre class="r"><code>library(cfdnakit)</code></pre>
<p>Let cfdnakit read sequence alignment file (.bam) with function read_bamfile. This function will split sequence reads into equal-size non-overlapping windows. Available size of bin are 100, 500, and 1000 KB.</p>
<pre class="r"><code>sample_bins = 
  read_bamfile(bamfile_path = &quot;/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/results_per_pid/OE0290-PED_5LB-017/alignment_umi/plasma-01-01_OE0290-PED_5LB-017_merged.mdup_groupbyUMI_callConsensus_realigned.bam&quot;,
               binsize = 1000)</code></pre>
<pre><code>## [1] &quot;Reading bamfile&quot;
## [1] &quot;Filtering-out read on blacklist&quot;</code></pre>
<pre class="r"><code>paste0(&quot;Number of bin : &quot;,length(sample_bins))</code></pre>
<pre><code>## [1] &quot;Number of bin : 2888&quot;</code></pre>
<p>Extract cfDNA fragment length information.</p>
<pre class="r"><code>sample_profile = 
  get_fragment_profile(sample_bins,
                       sample_id = &quot;plasma-01-01_OE0290-PED_5LB-017&quot;)
plot_isize_distribution(sample_profile)</code></pre>
<pre><code>## Warning: Removed 200 row(s) containing missing values (geom_path).</code></pre>
<p><img src="demonstration_markdown_files/figure-html/getting%20fragment%20length%20profile-1.png" width="672" /></p>
<pre class="r"><code>plot_sl_ratio(sample_profile)</code></pre>
<pre><code>## Warning: Removed 3 rows containing missing values (geom_point).</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_point).</code></pre>
<p><img src="demonstration_markdown_files/figure-html/getting%20fragment%20length%20profile-2.png" width="672" /></p>
<p>Save fragment profile as RDS file for later use or for creating Panel-of-Normal</p>
<pre class="r"><code>save_fragment_profile(sample_profile,
                      output_dir = &quot;/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/results_per_pid/OE0290-PED_5LB-017/cfdnakit/plasma-01-01_OE0290-PED_5LB-017/&quot;)</code></pre>
<pre><code>## [1] &quot;Saving RDS : Done&quot;</code></pre>
<p>Making a Panel-of-Normal is necessary for downstream analysis as we want to compare fragment profile between a cfDNA from patient with pooled of healthy individuals. First, we create a text file where each line is a full path to rds file created by aforementioned function.</p>
<pre class="r"><code>path_to_PoN_txt = &quot;/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/processing_scripts/cfdnakit_script/reference_healthy.listfile&quot;

create_PoN(path_to_PoN_txt,output_dir = &quot;/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/processing_scripts/cfdnakit_script/&quot;)</code></pre>
<pre><code>## [1] &quot;Done&quot;</code></pre>
<pre class="r"><code>PoN_rdsfile = &quot;/icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/processing_scripts/cfdnakit_script/PoN_fragment_profiles.rds&quot;
sample_zscore = get_zscore_profile(sample_profile,PoN_rdsfile)</code></pre>
<pre><code>## Reading PoN profile /icgc/dkfzlsdf/analysis/OE0290_projects/pediatric_tumor/whole_genome_sequencing/processing_scripts/cfdnakit_script/PoN_fragment_profiles.rds</code></pre>
<pre class="r"><code>sample_zscore_segment = segmentByPSCB(sample_zscore)</code></pre>
<pre><code>## Warning in storage.mode(x) &lt;- asMode: NAs introduced by coercion</code></pre>
<pre class="r"><code>plot_transformed_sl(sample_zscore,sample_zscore_segment)</code></pre>
<pre><code>## Warning: Removed 13 rows containing missing values (geom_point).</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_point).</code></pre>
<p><img src="demonstration_markdown_files/figure-html/Reading-in%20Panel-of-Normal%20samples%20and%20plot%20scaled%20value-1.png" width="672" /></p>
<pre class="r"><code>sample_cnv = call_cnv(sample_zscore_segment,sample_zscore)

plot_distance_matrix(sample_cnv)</code></pre>
<p><img src="demonstration_markdown_files/figure-html/cnv%20calling-1.png" width="672" /></p>
<pre class="r"><code>plot_cnv_solution(sample_cnv,selected_solution = 1)</code></pre>
<pre><code>## Warning: Removed 13 rows containing missing values (geom_point).</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_point).</code></pre>
<p><img src="demonstration_markdown_files/figure-html/cnv%20calling-2.png" width="672" /></p>
<pre class="r"><code>solution_table = get_solution_table(sample_cnv)</code></pre>
<pre class="r"><code>calculate_PCA_score(sample_zscore_segment)</code></pre>
<pre><code>## [1] 169.017</code></pre>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
